### ISSUE TITLE
Buffer limit of yamux is sometimes reached in substrate

### ISSUE THREAD

**Issue by tomaka** [libp2p/rust-libp2p#523]
Substrate seems to often reach the limit of the yamux internal buffer.

There should be a way in the API to increase the size, or make it infinite.

**Comment by twittner** [comment]
There is [`Config::set_max_buffer_size`](https://github.com/paritytech/yamux/blob/master/src/lib.rs#L74) to change the default. However the buffer size is *per stream*. Exceeding the buffer size means that the stream is not consuming incoming data (which arrives in 256 KiB frames). This seems to be a problem with task notification (see https://github.com/paritytech/yamux/pull/16).

**Comment by twittner** [comment]
Another possibility for the buffer overrun is that senders overwhelm receivers. Yamux implicitly sends window update frames when data frames are received, buffering up the data. If readers are slow consuming the buffer it will eventually reach its limit. The window update strategy was implemented this way to prevent deadlocks when both endpoints send data larger than the receiver's windows allow. If window updates would be sent only when data is consumed, both endpoints would be waiting simultaneously for window updates which never arrive if both do not read before finishing their writes.

The "on receive" strategy reduces the usefulness of window updates to help with head of line blocking but it does not really offer an effective back-pressure mechanism. If yamux users know their max. protocol unit size and configure their receive window accordingly and *always* interleave sends with reads, the "send window update only on read" strategy is more effective. I have opened https://github.com/paritytech/yamux/pull/17 to allow configuration of the window update mode.
